packages/mdc-textfield/input/foundation.js

```diff
@@ -0,0 +1,181 @@
+   * types.
+   * @return {!MDCTextFieldInputAdapter}
+   */
+  static get defaultAdapter() {
+    return /** @type {!MDCTextFieldInputAdapter} */ ({
+      registerEventHandler: () => {},
+      deregisterEventHandler: () => {},
+      getNativeInput: () => {},
+      notifyFocusAction: () => {},
+      notifyBlurAction: () => {},
+      notifyPressedAction: () => {},
+    });
+  }
+
+  /**
+   * @param {!MDCTextFieldInputAdapter=} adapter
+   */
+  constructor(adapter = /** @type {!MDCTextFieldInputAdapter} */ ({})) {
+    super(Object.assign(MDCTextFieldInputFoundation.defaultAdapter, adapter));
+
+    /** @private {boolean} */
+    this.receivedUserInput_ = false;
+    /** @private {function(): undefined} */
+    this.inputFocusHandler_ = () => this.activateFocus_();
+    /** @private {function(): undefined} */
+    this.inputBlurHandler_ = () => this.deactivateFocus_();
+    /** @private {function(): undefined} */
+    this.inputInputHandler_ = () => this.autoCompleteFocus_();
+    /** @private {function(): undefined} */
+    this.inputPressedHandler_ = () => this.notifyPressedAction_();
+  }
+
+  init() {
+    this.adapter_.registerEventHandler('focus', this.inputFocusHandler_);
+    this.adapter_.registerEventHandler('blur', this.inputBlurHandler_);
+    this.adapter_.registerEventHandler('input', this.inputInputHandler_);
+    ['mousedown', 'touchstart'].forEach((evtType) => {
+      this.adapter_.registerEventHandler(evtType, this.inputPressedHandler_);
+    });
+  }
+
+  destroy() {
+    this.adapter_.deregisterEventHandler('focus', this.inputFocusHandler_);
+    this.adapter_.deregisterEventHandler('blur', this.inputBlurHandler_);
+    this.adapter_.deregisterEventHandler('input', this.inputInputHandler_);
+    ['mousedown', 'touchstart'].forEach((evtType) => {
+      this.adapter_.deregisterEventHandler(evtType, this.inputPressedHandler_);
+    });
+  }
+
+  /**
+   * Activates the input's focus state.
+   * @private
+   */
+  activateFocus_() {
+    this.adapter_.notifyFocusAction();
+  }
+
+  /**
+   * Activates the input's focus state in cases when the input value
+   * changes without user input (e.g. programatically).
+   * @private
+   */
+  autoCompleteFocus_() {
+    if (!this.receivedUserInput_) {
+      this.activateFocus_();
+    }
+  }
+
+  /**
+   * Deactives the input's focus state.
+   * @private
+   */
+  deactivateFocus_() {
+    const input = this.getNativeInput_();
+
+    if (!input.value && !this.isBadInput()) {
+      this.receivedUserInput_ = false;
+    }
+
+    this.adapter_.notifyBlurAction();
+  }
+
+  notifyPressedAction_() {
+    this.adapter_.notifyPressedAction();
+  }
+
+  /**
+   * @return {boolean} True if the input fails validity checks.
+   */
+  isBadInput() {
+    const input = this.getNativeInput_();
+    return input.validity ? input.validity.badInput : input.badInput;
+  }
+
+  /**
+   * @return {string} Returns the value in the input.
+   */
+  getValue() {
```

Comment:
These APIs seem awkward for a foundation when they're just passing through a straight-up DOM operation. Is this something that would no longer be needed after #1675?

(I realize some of these e.g. isDisabled also existed prior to this PR...)

Reply:
The only other "non-awkward", option I can see is to have the foundation return `NativeInputType`. Like a `getNativeInputType()`. But I think I prefer we proxy the methods on `NativeInputType` as methods on the `MDCTextFieldInputFoundation` (aka .getValue and .isDisabled)

We would still need this after #1675. That would only change how we initiate these classes.